elytron/WFCORE-3947-ocsp-ssl-revocation.adoc

```diff
@@ -0,0 +1,75 @@
+= OCSP certificate revocation
+:author:            Martin Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3947[WFCORE-3947]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/ELY-1617[ELY-1617]
+
+=== Dev Contacts
+
+* mailto:mmazanek@redhat.com[Martin Mazanek]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* Security
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* New class RevocatingTrustManager
+* Configurable certificate revocation using OCSP. This will be implemented using PKIXRevocationChecker.
+* Configuration will be done using Builder class and will include:
+** Current configuration in X509CRLExtendedTrustManager
+** OCSP Responder URI for overriding default, where the default is retrieved from AIA extension (RFC5280) or from 'ocsp.responderURL' security property
+** Default revocation mechanism (OCSP/CRL) in a form of prefer crl (default false)
+** Fallback behaviour - cases for unavailable CRL / OCSP responder services
+** Checking of leaf certificates only
+* Deprecating X509CRLExtendedTrustManager, both mechanisms will share one class for fallback purposes and easier configuration
+
+Proposed configuration:
+
+```xml
+<trust-manager name="trust-with-ocsp-crl" algorithm="PKIX" key-store="ElytronCaTruststore">
+    <certificate-revocation crl-path="crl.pem" maximum-cert-path="2", key-store=trustedKS, soft-fail="true", prefer-crl="false", ocsp-responder="responderUrl"/>
+</trust-manager>
+```
+
+This would include both CRL and OCSP configuration in one thing, as they can't be configured separately due to sharing some parameters (eg. trusted key-store).
+Note: it is basically just refactoring and expanding existing CRL configuration, so we could stick to using element certificate-revocation-list for both CRL and OCSP to remain existing config,
+but I think refactoring it with just certificate-revocation will be less confusing. On top of that the behavior will slightly change, because certificate-revocation will enable OCSP by default.
+
+Other configuration is done globally on JVM scope, as documented in http://openjdk.java.net/jeps/249[JEP-249]
```

Comment:
It would make sense, but as it is not in JDK8 it could be confusing, as it would not work as configured. So I am not really sure with this one

Reply:
I think that in case we properly document this behaviour (in doc and also in an attribute/element description), it is not a problem. On JDK8 those values are simply ignored (optionally some warning may be generated for customer to be aware). There is already similar such thing for TLSv1.3 `cipher-suite-names` attribute - as TLSv1.3 is available since JDK-11 (well, TLSv1.3 will probably available with JDK8 in case when appropriate OpenSSL provider is used instead of default JSSE, but it's a different story, though).

Also, I can see that there is already this feature request https://issues.jboss.org/browse/ELY-434 to implement OCSP stampling support into WildFly. So maybe we can keep this proposal simply on a base OCSP support basis and 'official' stampling feature being implemented by another proposal - ELY-434. In such case this decision has to be stated in this analysis document. WDYT?

Reply:
This is mean as a base OCSP support RFE, stamping has its own issues. I would like to have them separate, as the stamping is JDK9+ and everything proposed here is in JDK8. The stamping support would consist mostly of tests and docs, as the configuration can be made through System and Security properties, which require no code change. So it will be possible to use this JKD8 targeted RFE with stamping released in JDK9 JEP249,